File: golang/modules/session.go
package modules

import (
	"errors"
	"strconv"
	"strings"

	"github.com/go-resty/resty/v2"
)

// SessionContext is the server-instance fencing tuple issued by NewNanManager.
type SessionContext struct {
	ID    string `json:"session_id"`
	Epoch int64  `json:"session_epoch"`
}

// Validate rejects a context that cannot represent an active server session.
func (session SessionContext) Validate() error {
	if strings.TrimSpace(session.ID) == "" {
		return errors.New("session id must not be blank")
	}
	if len(session.ID) < 32 || len(session.ID) > 64 {
		return errors.New("session id must be between 32 and 64 characters")
	}
	if session.Epoch <= 0 {
		return errors.New("session epoch must be greater than zero")
	}
	return nil
}

func applySessionHeaders(request *resty.Request, session SessionContext) error {
	if err := session.Validate(); err != nil {
		return err
	}
	request.SetHeader("X-NNM-Session-ID", session.ID)
	request.SetHeader("X-NNM-Session-Epoch", strconv.FormatInt(session.Epoch, 10))
	return nil
}

File: csharp/Models/Session.cs
using System.Text.Json.Serialization;

namespace NewNanManager.Client.Models;

/// <summary>
/// Server-instance session tuple issued by NewNanManager.
/// </summary>
public sealed class SessionContext
{
    public SessionContext(string id, long epoch)
    {
        if (string.IsNullOrWhiteSpace(id))
            throw new ArgumentException("Session id cannot be blank.", nameof(id));
        if (id.Length is < 32 or > 64)
            throw new ArgumentException("Session id length must be between 32 and 64 characters.", nameof(id));
        if (epoch <= 0)
            throw new ArgumentOutOfRangeException(nameof(epoch), "Session epoch must be greater than zero.");

        Id = id;
        Epoch = epoch;
    }

    /// <summary>Server-issued session identifier.</summary>
    public string Id { get; }

    /// <summary>Server-issued monotonic session epoch.</summary>
    public long Epoch { get; }

    internal IReadOnlyDictionary<string, string> Headers => new Dictionary<string, string>
    {
        ["X-NNM-Session-ID"] = Id,
        ["X-NNM-Session-Epoch"] = Epoch.ToString(System.Globalization.CultureInfo.InvariantCulture),
    };
}

/// <summary>
/// Wire response returned when a server session is created.
/// </summary>
public sealed class ServerSessionResponse
{
    [JsonPropertyName("session_id")]
    public string SessionId { get; set; } = string.Empty;

    [JsonPropertyName("session_epoch")]
    public long SessionEpoch { get; set; }

    internal SessionContext ToContext() => new(SessionId, SessionEpoch);
}
